### PR TITLE
IGNITE-22759 Do not do partition SafeTime sync if previous attempt is not finished

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-22759

### CODE REVIEW

**Review by vldpyatkov — APPROVED** [review]

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 1]
```diff
@@ -1086,16 +1097,64 @@ private void idleSafeTimeSync() {
...
 
-            replica.processRequest(req, localNodeId);
+        if (!shouldAdvanceIdleSafeTime(replica, proposedSafeTime)) {
+            // If previous attempt might still be waiting on the Metastorage SafeTime, we should not send the command ourselves as this
+            // might be an indicator that Metastorage SafeTime has stuck for some time; if we send the command, it will have to add its
+            // future, increasing (most probably, uselessly) heap pressure.
+            return;
         }
+
+        getOrCreateContext(replica).lastIdleSafeTimeProposal = proposedSafeTime;
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1114]: Do you expect proposedSafeTime to grow monotically?

File: modules/replicator/build.gradle [hunk 2]
```diff
@@ -37,6 +37,7 @@ dependencies {
     implementation project(':ignite-placement-driver-api')
     implementation project(':ignite-failure-handler')
     implementation project(':ignite-partition-distribution')
+    implementation project(':ignite-metastorage-api')
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/replicator/build.gradle:40]: It actually does not depend on the MG implementation, it's MG API I added a dependency from. Is this still bad?

Also, the idea of the fix is to look at ClusterTime (a Metastorage component), so replicator has to know about it. An alternative would be to do this check at the accepting side (`PartitionReplicaListener`), but with the current approach we save more resources (less objects are created). Or we could inverse dependencies by declaring an interface like `PrimaryInfo` (probably, a better name is to be invented) and move `shouldAdvanceIdleSafeTime()` method to it. Then implement the interface in the metastorage module and so on.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 3]
```diff
@@ -259,6 +264,7 @@ public ReplicaManager(
             RaftManager raftManager,
             RaftGroupOptionsConfigurer partitionRaftConfigurer,
             LogStorageFactoryCreator volatileLogStorageFactoryCreator,
+            ClusterTime clusterTime,
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:267]: Do we really need to write that `clusterTime is a cluster time`? :)

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 1 above)
Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1114]: Yes, as it's taken from the clock

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:267]: We should be consistent. Either all params should be mentioned or none.

File: modules/replicator/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/build.gradle:40]: > Is this still bad?

I guess so. Logically replicator should not depend on MG in any kind.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 1 above)
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1114]: In that case, seems that the only reason why we don't have races is the fact that scheduledIdleSafeTimeSyncExecutor is single treaded. It it's true, could you please add a comment somewhere near scheduledIdleSafeTimeSyncExecutor initialization denoting that it **should** be single threaded.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 4]
```diff
@@ -1086,16 +1097,64 @@ private void idleSafeTimeSync() {
...
+        getOrCreateContext(replica).lastIdleSafeTimeProposal = proposedSafeTime;
+
+        ReplicaSafeTimeSyncRequest req = REPLICA_MESSAGES_FACTORY.replicaSafeTimeSyncRequest()
+                .groupId(toReplicationGroupIdMessage(replica.groupId()))
+                .proposedSafeTime(proposedSafeTime)
+                .build();
+
+        replica.processRequest(req, localNodeId).whenComplete((res, ex) -> {
+            if (ex != null) {
+                LOG.error("Could not advance safe time for {} to {}", replica.groupId(), proposedSafeTime);
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1123]: Should we also log an exception? Debug will be difficult without that I guess.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 5]
```diff
@@ -1086,16 +1097,64 @@ private void idleSafeTimeSync() {
...
+            }
+        });
+    }
+
+    private boolean shouldAdvanceIdleSafeTime(Replica replica, HybridTimestamp proposedSafeTime) {
+        if (placementDriver.getCurrentPrimaryReplica(replica.groupId(), proposedSafeTime) != null) {
+            // We know a primary is guaranteed to be valid at the proposed safe time, so, when processing the SafeTime propagation request,
+            // we'll not need to wait for Metastorage SafeTime, so processing will be cheap.
+            return true;
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1132]: Curious, why not to awaite previous request to succeed if getPrimaryReplica(proceposedSafeTime) != null?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 6]
```diff
@@ -1086,16 +1097,64 @@ private void idleSafeTimeSync() {
...
+
+        replica.processRequest(req, localNodeId).whenComplete((res, ex) -> {
+            if (ex != null) {
+                LOG.error("Could not advance safe time for {} to {}", replica.groupId(), proposedSafeTime);
+            }
+        });
+    }
+
+    private boolean shouldAdvanceIdleSafeTime(Replica replica, HybridTimestamp proposedSafeTime) {
+        if (placementDriver.getCurrentPrimaryReplica(replica.groupId(), proposedSafeTime) != null) {
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1129]: I believe that we should move this check to the bottom and calculate proposed safeTime in front of it.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 7]
```diff
@@ -1086,16 +1097,64 @@ private void idleSafeTimeSync() {
...
+
+        // This is the Metastorage SafeTime that was needed to be achieved for previous attempt to check that this node is still a primary.
+        // If it's already achieved, then previous attempt is unblocked (most probably already finished), so we should proceed.
+        HybridTimestamp requiredMsSafeTime = lastIdleSafeTimeProposal.addPhysicalTime(clockService.maxClockSkewMillis());
+
+        return clusterTime.currentSafeTime().compareTo(requiredMsSafeTime) >= 0;
+    }
+
+    private ReplicaContext getOrCreateContext(Replica replica) {
+        return contexts.computeIfAbsent(replica.groupId(), k -> new ReplicaContext());
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1157]: Who is responsible for contexts map cleanup?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: This is not relevant anymore as the parameter is removed

File: modules/replicator/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/replicator/build.gradle]: I removed the dependency; instead, a method is added to `PlacementDriver`

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 5 above)
Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1126]: 1. If we await for the success of the previous attempt we might stuck forever if it never completes its future (for whatever reason). This was the main reason you rejected previous attempt to fix this problem in #4111 
2. The idea of this fix is to evade accumulating futures waiting on Metastorage SafeTime tracker. If `getCurrentPrimaryReplica()` returns a non-null value, we are sure there will no be wait even for THIS attempt, so we will not create a future. Other aspects don't bother us here (as they don't break correctness)

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 6 above)
Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1123]: What is your concern here? That now proposed safe time is generated once per batch (and you still want old behavior), or that `getCurrentPrimaryReplica()` for THIS attempt ts should be called and checked after checking that MS SafeTime is reached for the ts of the PREVIOUS attempt, or both?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 7 above)
Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1155]: Added removal of the context on replica stop

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 6 above)
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1126]: Every clock now() adds some contention on the clock. Thus if prev MG.safeTime await isn't finished yet there's no sense in calling clock.now().

Comment (rpuch) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1126]: We call `clock.now()` just once for the whole batch; so for each individual replica, `clock.now()` is not called, and there is no such calls that can be saved

Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:1126]: Up to you.

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriverManager.java [hunk 8]
```diff
@@ -278,6 +278,11 @@ private void recoverInternalComponentsBusy() {
 
     private PlacementDriver createPlacementDriver() {
         return new PlacementDriver() {
+            @Override
+            public boolean isActualAt(HybridTimestamp timestamp) {
+                return metastore.clusterTime().currentSafeTime().compareTo(timestamp) >= 0;
```
Comment (sanpwc) [inline modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriverManager.java:283]: Why there's no 'ts.addPhysicalTime(clockService.maxClockSkewMillis());'

Internally methods like getPrimaryReplica uses `clockService.after` that compares timestamps within the context of clock skew.
```
    default boolean after(HybridTimestamp firstTimestamp, HybridTimestamp anotherTimestamp) {
        return compareWithClockSkew(firstTimestamp, anotherTimestamp) > 0;
    }
```

Comment (rpuch) [reply modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/PlacementDriverManager.java:283]: Implementations of methods like `getAssignments()` do not add max clock skew when calculating the wait time. I suppose that a 'not adding' version is more universal as the caller can add the max clock skew (this is what I did: I add it externally to this method)

**Review by sanpwc — APPROVED** [review]

**Pull request merged**